### ISSUE TITLE
scatter_add allows broadcasting of index tensor in forward pass but crashes on backward pass

### ISSUE THREAD

**Issue by HapeMask** [pytorch/pytorch#27614]
## 🐛 Bug

If you have an index tensor whose shape does not match the other tensor's shape but is broadcast-compatible, the forward pass of `torch.scatter_add()` will succeed but the backward pass will throw a RuntimeError saying that the shapes do not match exactly on the non-indexed dimension.

## To Reproduce

Repro script:
```python
import torch as th

src = th.zeros(8, 105238, 3)
ind = th.zeros(8, 626304, 1).long()
ofs = th.zeros(8, 626304, 3)
src.requires_grad_(True)
ofs.requires_grad_(True)

out = src.scatter_add(1, ind, ofs)
#out = src.scatter_add(1, ind.expand(-1, -1, 3), ofs) <<< this line works

l = out.sum()
l.backward()
```

Exception (CPU):
```
RuntimeError: Expected tensor [8, 626304, 1], src [8, 105238, 3] and index [8, 626304, 1] to have the same size ap
art from dimension 1
```

Exception (CUDA):
```
RuntimeError: invalid argument 2: Input tensor must have same size as output tensor apart from the specified dimen
sion at /mnt/home/gbschwartz/pytorch/aten/src/THC/generic/THCTensorScatterGather.cu:27
```

## Expected behavior

No exceptions thrown.

## Environment

 - PyTorch Version (e.g., 1.0): 1.2.0
 - OS (e.g., Linux): linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 9.0 / 7.6.0
 - GPU models and configuration: V100

**Comment by zou3519** [comment]
Crashes are bad, we should fix this.

**Comment by gchanan** [comment]
this doesn't crash, it throws an error.

**Comment by gchanan** [comment]
I also don't think this is related to broadcasting.  The same example works in the forward pass if you replace the "1" with a "2", which isn't broadcast compatible.

We've had various consistency issues with what scatter/gather functions supported in CPU/CUDA forward/backward.  We should do a audit of this code across those axes.

**Comment by gchanan** [comment]
@ifedan is going to take a look.

**Comment by gchanan** [comment]
ping @ifedan -- any update?

**Comment by rgommers** [comment]
@nikitaved will start working on this (can't assign him yet, needs adding to a team)

**Comment by nikitaved** [comment]
Yes, I can have a look. However, I am not super familiar with the codebase yet. Could you please hint me where it is the best to first look at?

**Comment by nikitaved** [comment]
Hi! So, it seems to me I know where the issue comes from, for CPU at least...

So, for the forward pass `TH_TENSOR_DIM_APPLY3_SCATTER` macro is used for the dimensionaliry check https://github.com/pytorch/pytorch/blob/b44657299705ccb89ba84427911fc3d5d3b17ebf/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L692
The macro is defined in `aten/src/TH/generic/THTensorApply.hpp` and has the following description:

```
// Used for scatter and scatterAdd 
// Assumes TENSOR1 is index  
//         TENSOR2 is real  
//         TENSOR3 is src  
// Tests:  
//   1. index->size(d) <= src->size(d) for all d  
//   2. index->size(d) <= real->size(d) for all d != dim
```  

So, this is why the forward pass does not fail.

Now, in the backward pass `scatter_add` uses `gather`, which can be seen in `tools/autograd/derivatives.yml`. `gather` is also defined in `aten/src/TH/generic/THTensorEvenMoreMath.cpp`, but for the dimensionality check it uses `TH_TENSOR_DIM_APPLY3_SIZE_EQ_EXCEPT_DIM` macros. The macros checks for the exact dimension match but a selected dimension, and **this is where the issue triggers**. I did not see any broadcasting involved however...

@gchanan , do you think it is safe to just replace `TH_TENSOR_DIM_APPLY3_SIZE_EQ_EXCEPT_DIM` with
`TH_TENSOR_DIM_APPLY3_SCATTER` in the `gather` method?
https://github.com/pytorch/pytorch/blob/b44657299705ccb89ba84427911fc3d5d3b17ebf/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L633
Could you also please tell me where in the code the broadcasting takes place?

Thanks!

**Comment by nikitaved** [comment]
@gchanan, @ezyang would just changing the assertion do for now? Maybe it would be also nice to port the legacy code?

**Comment by ezyang** [comment]
I believe the broadcasting occurs in the backwards of sum: if the inflowing gradient is `x`, then the outflowing gradient is `x.expand_as(input.size())`

**Comment by nikitaved** [comment]
Ok, but here the issue is with the index tensor in forward/backward pass. Following the example in the repro script, we see that `scatter_add` produces an output of dimension `[8, 626304, 3]`. Hence the inflowing gradient `grad` would have the same dimension. The backward method for `scatter_add` is defined as `grad.gather(dim, index)', and produces an output of the same size as `index`, which is `[8, 626304, 1]`. So, there is a dimension mismatch between forward and backward.

There is also, I believe, an issue in the documentation. It says that `gather` and `scatter` are sort of 'inverse' of each other, but this is wrong, because the shape of the output of `gather' does depend on the shape of index, and this is not the case for `scatter`.

**Comment by albanD** [comment]
@nikitaved does `grad.gather(dim, index.expand_as(src))` be a correct formula in this case?

**Comment by nikitaved** [comment]
@albanD, no, it is not, not in the way that `scatter_add` is implemented. In `scatter_add` it hold that `index.shape <= src.shape'. The original formula is correct, and so is yours, when `index.shape = src.shape'.

**Comment by albanD** [comment]
My logic is as follows. Can you let me know why it is wrong?

But `scatter_add` requires `index` and `src` to have the same number of dimensions.
It also requires for each dim that the size of index <= size of src.
And broadcasting in this case boils down to `index.expand_as(src)` right?
So adding the expand explicitly in the formula will give a gradient for `src` that is the same size as `src` even when `index` was broadcasted.

**Comment by nikitaved** [comment]
Let
```latex
     I = {(i_1,...,i_n) | (i_1,...,i_n) < index.shape},
     U = {(i_1,...,i_n) | (i_1,...,i_n) < source.shape},
 co(I) = U/I,
```
where `<` is element-wise. Now, from the definition of `scatter_add` we can see that `source` contributes only elements `source[I]` to the output, which means that if `grad` is the inflow grad into `scatter_add`, only `grad[I]` will flow back into `source[I]`, then the right gradient wrt `source`, `grad_source` is
```
grad_source[I] = grad.gather(dim, index),
grad_source[co(I)] = 0.
```
Please, let me know if it is correct.

**Comment by ezyang** [comment]
Thanks, my earlier comment was wrong. Sorry about the confusion.

**Comment by ezyang** [comment]
> otherwise gather has no idea about the shape of source when we want to backprop to it, and then we could leave backward for scatter_add as it is.

I'm still reading the rest of the comments and correcting my misunderstanding, but in this particular case, the way you solve it is by having the backwards formula save the shape of the original input, so you know how big to make `grad_input` the other way around. This is for example, how convolution backwards are implemented, as rounding means the input size is not uniquely determined by the backward "inputs".

**Comment by ezyang** [comment]
@nikitaved I agree with your reasoning. As @gchanan originally pointed out, this issue has nothing to do with broadcasting. This example makes the behavior clear:

```
>>> x
tensor([[0.4633, 0.9984, 0.9264, 0.7699, 0.8603],
        [0.0634, 0.8075, 0.1522, 0.9515, 0.4617]])
>>> torch.ones(3, 5).scatter_add(0, torch.tensor([[0,0,0,0,0], [2,2,2,2,2]]), x)tensor([[1.4633, 1.9984, 1.9264, 1.7699, 1.8603],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0634, 1.8075, 1.1522, 1.9515, 1.4617]])
>>> torch.ones(3, 5).scatter_add(0, torch.tensor([[0,0,0,0,0]]), x)
tensor([[1.4633, 1.9984, 1.9264, 1.7699, 1.8603],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000]])
```

**Comment by albanD** [comment]
Ok, after some testing. My confusion comes from the fact that I assumed that: `out = src.scatter_add(1, ind, ofs)` and `out = src.scatter_add(1, ind.expand(-1, -1, 3), ofs)` were the same thing. But they are not.
I was expecting the indices to be expanded under the hood.

I would argue that this behavior is a bug and should not be allowed?
The fix for existing code is simple: narrow the output so that it has the same size as the indices.

**Comment by ezyang** [comment]
@albanD It's not that simple. You have to do a multi-dimension narrow:

```
>>> x = torch.rand(3, 2, 5)
>>> torch.ones(3, 3, 5).scatter_add(1, torch.tensor([[[0,0,0,0], [1,1,1,1]]]), x)
tensor([[[1.8910, 1.2403, 1.2806, 1.4406, 1.0000],
         [1.1750, 1.1589, 1.4555, 1.7350, 1.0000],
         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000]],

        [[1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000]],

        [[1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000]]])
>>> torch.tensor([[[0,0,0,0], [1,1,1,1]]]).size()
(1, 2, 4)
```

Multi-dim narrow is possible but I don't think we have any function that does it. Also, to expand on the suggestion, the claim is to make `scatter_add` a composite operation which does a narrow and then a `_scatter_add`, where the latter requires input and index to match exactly. Then the existing backwards formula for narrow will handle filling in the zeros correctly. I suppose you could also just write a fused backwards that does it all in one go too; might be marginally faster.

**Comment by albanD** [comment]
> Multi-dim narrow is possible but I don't think we have any function that does it.

Shall we add a `narrow_as` ? 

> Also, to expand on the suggestion, the claim is to make scatter_add a composite operation

My suggestion was not to make `scatter_add` a composite function but make a BC-breaking change where we only allow indices and source to match exactly. The message corresponding to this error would be `Careful, part of your source will be ignored. If you really want this, narrow your source to match your indices before calling scatter_add`.
This could be generalized to all the `gather*` functions.

**Comment by gchanan** [comment]
> My suggestion was not to make scatter_add a composite function but make a BC-breaking change where we only allow indices and source to match exactly. The message corresponding to this error would be Careful, part of your source will be ignored. If you really want this, narrow your source to match your indices before calling scatter_add.

Thus my suggestion above to do an audit of all this code.  What constraints do scatter/scatter_add/gather/etc impose across CPU/CUDA/forward/backward boundaries?

**Comment by nikitaved** [comment]
@ezyang, so, currently there is no functional to "subtenso"r `source` to match the size of `index`?

**Comment by ezyang** [comment]
@nikitaved I'm not familiar with any such function. In particular, there's a limited number of view functions and I don't think any of them fit the bill here.

**Comment by nikitaved** [comment]
All right, then indeed it seems like a great opportunity to actually port scatter/gather functions to ATen. If it is fine by you, I could give it a try.

**Comment by ezyang** [comment]
SGTM. Pull up the relevant issue and assign yourself to it.

**Comment by nikitaved** [comment]
@gchanan ,
> We've had various consistency issues with what scatter/gather functions supported in CPU/CUDA forward/backward. We should do a audit of this code across those axes.

Could you please give some examples? Or do you think they are all related to this issue?

**Comment by nikitaved** [comment]
What about actually modifying `gather` method to return output of any specified shape greater than `index.shape`, with default value equals `index.shape`? There is no need to modify c++ for that, and for the `scatter_add.backward` we can record `src.shape`.

**Comment by ezyang** [comment]
@nikitaved I think what @gchanan is suggesting, is before we talk about implementation strategies for how to fix this, we need to work out what the desired semantics of `scatter_add` are. But to answer that question, we have to look at how some of the closely related functions currently behave, and see if there is a consistent rule we can apply across all of them. I can't really answer your latest question without an understanding about what the planned semantics are.

**Comment by nikitaved** [comment]
Upon further review, the proposed modified backward formula is only correct if `index` has only unique indices! The documentation for `scatter/gather` says that `index` has to contain unique indices, but, in practice, this is not the case, as was pointed out in [#31662](https://github.com/pytorch/pytorch/pull/31662). The backward methods for `scatter/gather` have to be modified (see discussion in [#31662](https://github.com/pytorch/pytorch/pull/31662)).

**Comment by yael-ai** [comment]
Hello
I am using the .gather function in my loss function an I get this error: 
RuntimeError: Expected tensor [32, 1], src [128, 5] and index [32, 1] to have the same size apart from dimension 1
 I am running this:
state_action_values = state_action_values.gather(1,batch.action.unsqueeze(1)).squeeze(1)
the shape of state_action_values is[128,5]
and the shape of batch.action.unsqueeze(1) is [32,1]
I do not understand where is the problem, what dimensions should I have?
